### ISSUE TITLE
Covering binaries on OS X

### ISSUE THREAD

**Issue by zbeekman** [SimonKagstrom/kcov#135]
Will this build and install on OS X? I'm guessing the answer is "no" because ELF/DWARF is Linux specific? (Please forgive my crude understanding of everything deeper than compiled language source code... I am but a lowly aerospace engineer/computational scientist with no formal CS background)

**Comment by SimonKagstrom** [comment]
Technically, the answer is "yes" :-). However, on OS X, only Python and Bash coverage collection works. And as you guess, the ELF/ptrace parts are Linux-specific. DWARF is used on OS X as well, so a port is certainly possible, but will require a bit of work:

- kcov uses libdw from elfutils for the DWARF parsing, which is closely tied to ELF
- the ptrace parts will be similar, but the interface is not exactly the same, so this will have to be split
- a macho-parser.cc (how's that for a name!) that implements the needed parts from elf-parser.cc will have to be created

Unfortunately, I don't run OS X myself (although I use Apple hardware, but with Linux), so I don't really have a time frame for a full port to OS X.

**Comment by zbeekman** [comment]
Great thanks for clarifying!

One more follow up question: If I only want to generate shell script coverage data, are there features in kcov that are missing from shcov? It seems that building kcov is a little bit more involved, and I was wondering if the feature set for shell script coverage info is the same between the two.

Not having to compile with gcov flags is an awesome feature of kcov, so I was considering dropping the "Coverage" build from our Travis-CI test matrix and using kcov to get coverage for the release builds. It sounds like I can do this on the Linux VMs but the OS X VMs will still need to  use gcov for coverage data.

**Comment by zbeekman** [comment]
P.S. thanks for the cool project!

**Comment by SimonKagstrom** [comment]
Thanks!

Yes, shcov is quite a bit more limited than kcov: It doesn't support coveralls, has more basic HTML output, can't accumulate data between runs, no filtering and a bit more false positives/negatives for bash.

I'll keep this bug report open for the OS X collection, but don't expect to get much work done on it for a while!

**Comment by SimonKagstrom** [comment]
By the way: If you're using Travis-CI for Bash collection on OS X, the travis/Makefile file in the kcov distribution contains the needed homebrew packages for the kcov build:

```
brew update
brew install homebrew/dupes/zlib
brew install python
brew install bash
```

**Comment by zbeekman** [comment]
Thanks. So I take it that kcov requires bash 4?

**Comment by SimonKagstrom** [comment]
No, Bash 4.2 (I think it was) adds some support for tracing to be done without PS4. With older bashes, it will still work, falling back to PS4.

There are some special options to control this behavior, and things like detecting #!/bin/sh invocations where /bin/sh is not bash, but those doesn't work everywhere so are off by default.

**Comment by zbeekman** [comment]
@SimonKagstrom I tried installation on OS X but CMake barfs at Line 13 in src/CMakeLists.txt:

```
pkg_check_modules(LIBZ REQUIRED zlib)
```

I installed homebrew/dupes/zlib as you suggested, but this still fails.

Also, could you please explain the usage of `pkg_check_modules` over `find_package`? I'm at least moderately familiar with CMake but am unfamiliar with `pkg_check_modules`

**Comment by SimonKagstrom** [comment]
OK, unfortunately I've only tried the OS X stuff in Travis, where I only do the commands above. I guess Travis might pre-install some packages on OS X as well though, which might explain why it works for me.

pkg-config is a tool and "database" for libraries. You can use it to find headers and linking options provided by just the library name, e.g.,

```
$ pkg-config --libs zlib
-L/usr/lib64 -lz
```

and with cmake, this means that you don't have to have/write a find-package file for each library, and instead you can use the generic pkg_check_modules. Perhaps there is some (to me) unknown dependency for cmake + pkg-config which causes the problem?

**Comment by luser** [comment]
Google Breakpad has some BSD-licensed Mach-O parsing code if anyone is interested in trying this:
https://chromium.googlesource.com/breakpad/breakpad/+/master/src/common/mac/macho_reader.cc

**Comment by SimonKagstrom** [comment]
OK, interesting, thanks!

I guess I should be getting a machine to run Mac OS to really be able to work on this though.

**Comment by SimonKagstrom** [comment]
I've looked a bit more on this after buying myself an IMac :-)

And what I think needs to be done, apart from ptrace, is to be able to read the dSYM bundles / files. As far as I understand, these are pretty much raw DWARF files, but I've yet to find a library that manages to read  them. libdwarf (through dwarfutils) is available on OSX homebrew, but as far as I can see is really tied to the ELF format - dwarfdump also cannot read dSYM files.

So this is my current hurdle, not sure how to proceed there.

**Comment by luser** [comment]
Breakpad has a full set of DWARF reading code (see dwarf_* and the files under dwarf/):
https://chromium.googlesource.com/breakpad/breakpad/+/master/src/common/

We use that on Linux and OS X. For OS X we use the macho_reader code I linked above to parse the Mach-O binary format and then feed the debug sections to the DWARF parser. Here's the bulk of the code that drives that:
https://chromium.googlesource.com/breakpad/breakpad/+/master/src/common/mac/dump_syms.cc

And here's the entry point of the `dump_syms` tool that uses that code:
https://chromium.googlesource.com/breakpad/breakpad/+/master/src/tools/mac/dump_syms/dump_syms_tool.cc

The `dump_syms` tool reads function + line + CFI data out of binaries and prints it in a text format that Breakpad uses for unwinding stacks from crash dumps.

**Comment by SimonKagstrom** [comment]
Thanks for pointing me in the right direction Ted! Will definitely play around with that!

Another idea I had was otherwise to use the lldb library which should have the line/file/address mappings as well, and can also obviously set and clear breakpoints. However, Apple doesn't appear to ship the header file, so I guess that will be difficult anyway.

**Comment by luser** [comment]
The LLDB library would be a pretty good choice, but Apple not shipping the headers makes that kind of annoying. It looks like you could plausibly work around that by doing something like this project has:
https://github.com/eonil/LLDBWrapper#caution

**Comment by SimonKagstrom** [comment]
Yes, I think the LLDB approach looks promising. I've been experimenting a bit with it, and I'm able to read the file:line -> address mappings so far, so I think it should be workable and fairly easy.

I looked at how this project used the lldb library, https://github.com/freedib/lldbmi2, and with a few CMakeLists.txt changes, I'm able to compile and link against the LLDB library.

**Comment by SimonKagstrom** [comment]
Fixed with 53eae50.

However, see Issues #157 #158 and #159 for a few limitations and bugs (to be fixed). The implementation is also extremely slow, so probably not very useful for larger programs right now.
